src/factorizations.jl

```diff
@@ -1,45 +1,83 @@
+  return qr(A, biperm)
+end
+function qr(A::AbstractArray, biperm::BlockedPermutation{2}; full::Bool=true, kwargs...)
+  # tensor to matrix
+  A_mat = fusedims(A, biperm)
+
+  # factorization
+  Q, R = full ? qr_full(A_mat; kwargs...) : qr_compact(A_mat; kwargs...)
+
+  # matrix to tensor
+  axes_codomain, axes_domain = blockpermute(axes(A), biperm)
+  axes_Q = (axes_codomain..., axes(Q, 2))
+  axes_R = (axes(R, 1), axes_domain...)
+  return splitdims(Q, axes_Q), splitdims(R, axes_R)
+end
+
+"""
+    svd(A::AbstractArray, labels_A, labels_codomain, labels_domain; full=false, kwargs...)
+    svd(A::AbstractArray, biperm::BlockedPermutation{2}; full=false, kwargs...)
+
+Compute the SVD decomposition of a generic N-dimensional array, by interpreting it as
+a linear map from the domain to the codomain indices. These can be specified either via
+their labels, or directly through a `biperm`.
+"""
+function svd(A::AbstractArray, labels_A, labels_codomain, labels_domain; kwargs...)
+  biperm = blockedperm_indexin(Tuple.((labels_A, labels_codomain, labels_domain))...)
+  return svd(A, biperm; kwargs...)
 end
+function svd(A::AbstractArray, biperm::BlockedPermutation{2}; full::Bool=false, kwargs...)
+  # tensor to matrix
+  A_mat = fusedims(A, biperm)
+
+  # factorization
+  U, S, Vᴴ = full ? svd_full(A_mat; kwargs...) : svd_compact(A_mat; kwargs...)
 
-function qr(a::AbstractArray, labels_a, labels_codomain, labels_domain)
-  # TODO: Generalize to conversion to `Tuple` isn't needed.
-  return qr(
-    a, blockedperm_indexin(Tuple(labels_a), Tuple(labels_codomain), Tuple(labels_domain))
-  )
+  # matrix to tensor
+  axes_codomain, axes_domain = blockpermute(axes(A), biperm)
+  axes_U = (axes_codomain..., axes(U, 2))
+  axes_Vᴴ = (axes(Vᴴ, 1), axes_domain...)
+  return splitdims(U, axes_U), S, splitdims(Vᴴ, axes_Vᴴ)
 end
 
-function svd(a::AbstractArray, biperm::BlockedPermutation{2})
-  a_matricized = fusedims(a, biperm)
-  usv_matricized = LinearAlgebra.svd(a_matricized)
-  u_matricized = usv_matricized.U
-  s_diag = usv_matricized.S
-  v_matricized = usv_matricized.Vt
-  axes_codomain, axes_domain = blockpermute(axes(a), biperm)
-  axes_u = (axes_codomain..., axes(u_matricized, 2))
-  axes_v = (axes(v_matricized, 1), axes_domain...)
-  u = splitdims(u_matricized, axes_u)
-  # TODO: Use `DiagonalArrays.diagonal` to make it more general.
-  s = Diagonal(s_diag)
-  v = splitdims(v_matricized, axes_v)
-  return u, s, v
+# TODO: decide on interface
+"""
+    tsvd(A::AbstractArray, labels_A, labels_codomain, labels_domain; kwargs...)
+    tsvd(A::AbstractArray, biperm::BlockedPermutation{2}; kwargs...)
+
+Compute the truncated SVD decomposition of a generic N-dimensional array, by interpreting it
+as a linear map from the domain to the codomain indices. These can be specified either via
+their labels, or directly through a `biperm`.
+"""
+function tsvd(A::AbstractArray, labels_A, labels_codomain, labels_domain; kwargs...)
```

Comment:
Just a suggestion, but what if truncated SVD is implemented as an algorithm backend of a more general `svd` function?

Reply:
I'm definitely also fine with that, behind the scenes this is already what is going on in MatrixAlgebraKit as well, see the definition [here](https://github.com/QuantumKitHub/MatrixAlgebraKit.jl/blob/14ed4c39f6d7911cd1c98658e7535edfa70d52fc/src/implementations/truncation.jl#L75). It's just a matter of deciding on where you want to start intercepting the user-input and what you would like the keywords for that to be, and I'm happy to map it to whatever MatrixAlgebraKit has

Reply:
Yeah, that is definitely a bit subtle, since ITensor has its own syntax for truncation keyword arguments.

Maybe a strategy we can take for now is starting out by following whatever MatrixAlgebraKit.jl does (say `TensorAlgebra.svd` just forwards to `MatrixAlgebraKit.svd`), and then at the ITensor level we can map our truncation arguments like `cutoff` and `maxdim` to the TensorAlgebra.jl/MatrixAlgebraKit.jl names and conventions, and additionally choose `TruncatedAlgorithm` as the default SVD algorithm.